File: slicing-ios/SlicingGame/Components/BasicViews/ImageButtonView.swift
//
//  ImageButtonView.swift
//  Basic view: an image which can be used as a button
//

import UIKit
import UniLayout
import JsonInflator

class ImageButtonView: UniImageView, ControlComponent {
    
    // --
    // MARK: Members
    // --
    
    var isEnabled: Bool {
        set {
            isUserInteractionEnabled = newValue
        }
        get { return isUserInteractionEnabled }
    }

    var isHighlighted = false {
        didSet {
            updateState()
        }
    }

    private let touchInsideTolerance: CGFloat = 64
    private var stateImage = ComponentStateImage()

    
    // --
    // MARK: Viewlet integration
    // --

    class func viewlet() -> JsonInflatable {
        return ViewletClass()
    }
    
    private class ViewletClass: JsonInflatable {
        
        func create() -> Any {
            return ImageButtonView()
        }
        
        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            if let imageButton = object as? ImageButtonView {
                // Apply images
                let stretchType = ImageStretchType(rawValue: convUtil.asString(value: attributes["stretchType"]) ?? "") ?? .none
                imageButton.internalImageView.contentMode = stretchType.toContentMode()
                imageButton.source = ImageSource.fromValue(value: attributes["source"])
                imageButton.highlightedSource = ImageSource.fromValue(value: attributes["highlightedSource"])
                imageButton.disabledSource = ImageSource.fromValue(value: attributes["disabledSource"])
                
                // Apply colorization
                imageButton.colorize = convUtil.asColor(value: attributes["colorize"])
                imageButton.highlightedColorize = convUtil.asColor(value: attributes["highlightedColorize"])
                imageButton.disabledColorize = convUtil.asColor(value: attributes["disabledColorize"])

                // Generic view properties
                ViewletUtil.applyGenericViewAttributes(convUtil: convUtil, view: imageButton, attributes: attributes)

                // Apply event
                imageButton.tapEvent = AppEvent.fromValue(value: attributes["tapEvent"])

                // Chain event observer
                if let eventObserver = parent as? AppEventObserver {
                    imageButton.eventObserver = eventObserver
                }
                return true
            }
            return false
        }
        
        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return type(of: object) == ImageButtonView.self
        }
        
    }
    
    
    // --
    // MARK: Initialization
    // --
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }
    
    private func setup() {
        clipsToBounds = true
        isUserInteractionEnabled = false
        tintAdjustmentMode = .normal
        internalImageView.contentMode = .center
    }
    
    
    // --
    // MARK: Configurable values
    // --
    
    weak var eventObserver: AppEventObserver?
    
    var tapEvent: AppEvent? {
        didSet {
            isEnabled = tapEvent != nil
        }
    }

    var source: ImageSource? {
        didSet {
            if source !== oldValue {
                if let source = source {
                    source.getImage(completion: { [weak self] image in
                        if self?.source === source {
                            self?.stateImage.image = image
                            self?.updateState()
                        }
                    })
                } else {
                    stateImage.image = nil
                    updateState()
                }
            }
        }
    }

    var highlightedSource: ImageSource? {
        didSet {
            if highlightedSource !== oldValue {
                if let highlightedSource = highlightedSource {
                    highlightedSource.getImage(completion: { [weak self] image in
                        if self?.highlightedSource === highlightedSource {
                            self?.stateImage.highlightedImage = image
                            self?.updateState()
                        }
                    })
                } else {
                    stateImage.highlightedImage = nil
                    updateState()
                }
            }
        }
    }

    var disabledSource: ImageSource? {
        didSet {
            if disabledSource !== oldValue {
                if let disabledSource = disabledSource {
                    disabledSource.getImage(completion: { [weak self] image in
                        if self?.disabledSource === disabledSource {
                            self?.stateImage.disabledImage = image
                            self?.updateState()
                        }
                    })
                } else {
                    stateImage.disabledImage = nil
                    updateState()
                }
            }
        }
    }
    
    var colorize: UIColor? {
        set {
            stateImage.colorize = newValue
            updateState()
        }
        get { return stateImage.colorize }
    }

    var highlightedColorize: UIColor? {
        set {
            stateImage.highlightedColorize = newValue
            updateState()
        }
        get { return stateImage.highlightedColorize }
    }

    var disabledColorize: UIColor? {
        set {
            stateImage.disabledColorize = newValue
            updateState()
        }
        get { return stateImage.disabledColorize }
    }

    
    // --
    // MARK: Interaction
    // --

    open override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        if let _ = touches.first {
            isHighlighted = true
            return
        }
        super.touchesBegan(touches, with: event)
    }
    
    open override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        if let touch = touches.first {
            let position = touch.location(in: self)
            isHighlighted = position.x >= -touchInsideTolerance && position.x < frame.width + touchInsideTolerance && position.y >= -touchInsideTolerance && position.y < frame.height + touchInsideTolerance
            return
        }
        super.touchesMoved(touches, with: event)
    }

    open override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        if let touch = touches.first {
            isHighlighted = false
            let position = touch.location(in: self)
            if position.x >= -touchInsideTolerance && position.x < frame.width + touchInsideTolerance && position.y >= -touchInsideTolerance && position.y < frame.height + touchInsideTolerance {
                buttonTapped()
            }
            return
        }
        super.touchesEnded(touches, with: event)
    }
    
    private func buttonTapped() {
        if let tapEvent = tapEvent {
            eventObserver?.observedEvent(tapEvent, sender: self)
        }
    }

    
    // --
    // MARK: State handling
    // --
    
    override var isUserInteractionEnabled: Bool {
        didSet {
            updateState()
        }
    }
    
    private func updateState() {
        if isEnabled {
            stateImage.state = isHighlighted ? .highlighted : .normal
        } else {
            stateImage.state = .disabled
        }
        image = stateImage.currentImage()
        internalImageView.tintColor = stateImage.currentColorize()
    }
    
}

File: slicing-android/app/src/main/java/com/crescentflare/slicinggame/components/basicviews/ImageButtonView.kt
package com.crescentflare.slicinggame.components.basicviews

import android.annotation.TargetApi
import android.content.Context
import android.graphics.Color
import android.os.Build
import android.util.AttributeSet
import android.view.View
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil
import com.crescentflare.slicinggame.components.utility.ComponentStateDrawable
import com.crescentflare.slicinggame.components.utility.ImageSource
import com.crescentflare.slicinggame.components.utility.ViewletUtil
import com.crescentflare.slicinggame.infrastructure.events.AppEvent
import com.crescentflare.slicinggame.infrastructure.events.AppEventObserver
import com.crescentflare.unilayout.views.UniImageView
import java.lang.ref.WeakReference


/**
 * Basic view: an image which can be used as a button
 */
open class ImageButtonView : UniImageView, View.OnClickListener {

    // --
    // Static: viewlet integration
    // --

    companion object {

        val viewlet: JsonInflatable = object : JsonInflatable {
            override fun create(context: Context): Any {
                return ImageButtonView(context)
            }

            override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
                if (obj is ImageButtonView) {
                    // Apply image
                    val stretchType = ImageView.StretchType.fromString(mapUtil.optionalString(attributes, "stretchType", ""))
                    obj.scaleType = stretchType.toScaleType()
                    obj.source = ImageSource.fromValue(attributes["source"])
                    obj.highlightedSource = ImageSource.fromValue(attributes["highlightedSource"])
                    obj.disabledSource = ImageSource.fromValue(attributes["disabledSource"])

                    // Apply colorization
                    obj.colorize = mapUtil.optionalColor(attributes, "colorize", Color.TRANSPARENT)
                    obj.highlightedColorize = mapUtil.optionalColor(attributes, "highlightedColorize", Color.TRANSPARENT)
                    obj.disabledColorize = mapUtil.optionalColor(attributes, "disabledColorize", Color.TRANSPARENT)

                    // Generic view properties
                    ViewletUtil.applyGenericViewAttributes(mapUtil, obj, attributes)

                    // Apply event
                    obj.tapEvent = AppEvent.fromValue(attributes["tapEvent"])

                    // Chain event observer
                    if (parent is AppEventObserver) {
                        obj.eventObserver = parent
                    }
                    return true
                }
                return false
            }

            override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
                return obj::class == ImageButtonView::class
            }
        }
    }


    // --
    // Members
    // --

    private val stateDrawable = ComponentStateDrawable()
    private var eventObserverReference : WeakReference<AppEventObserver>? = null


    // --
    // Initialization
    // --

    @JvmOverloads
    constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0)
            : super(context, attrs, defStyleAttr)

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int)
            : super(context, attrs, defStyleAttr, defStyleRes)

    init {
        setImageDrawable(stateDrawable)
    }


    // --
    // Configurable values
    // --

    var eventObserver: AppEventObserver?
        get() = eventObserverReference?.get()
        set(newValue) {
            eventObserverReference = if (newValue != null) {
                WeakReference(newValue)
            } else {
                null
            }
        }

    var tapEvent: AppEvent? = null
        set(tapEvent) {
            field = tapEvent
            setOnClickListener(if (tapEvent != null) this else null)
        }

    var source: ImageSource? = null
        set(source) {
            if (field !== source) {
                field = source
                if (source != null) {
                    source.getDrawable(context) {
                        if (field === source) {
                            stateDrawable.drawable = it
                        }
                    }
                } else {
                    stateDrawable.drawable = null
                }
            }
        }

    var highlightedSource: ImageSource? = null
        set(highlightedSource) {
            if (field !== highlightedSource) {
                field = highlightedSource
                if (highlightedSource != null) {
                    highlightedSource.getDrawable(context) {
                        if (field === highlightedSource) {
                            stateDrawable.pressedDrawable = it
                        }
                    }
                } else {
                    stateDrawable.pressedDrawable = null
                }
            }
        }

    var disabledSource: ImageSource? = null
        set(disabledSource) {
            if (field !== disabledSource) {
                field = disabledSource
                if (disabledSource != null) {
                    disabledSource.getDrawable(context) {
                        if (field === disabledSource) {
                            stateDrawable.disabledDrawable = it
                        }
                    }
                } else {
                    stateDrawable.disabledDrawable = null
                }
            }
        }

    var colorize: Int?
        set(colorize) {
            stateDrawable.colorize = if (colorize == Color.TRANSPARENT) null else colorize
        }
        get() = stateDrawable.colorize

    var highlightedColorize: Int?
        set(highlightedColorize) {
            stateDrawable.pressedColorize = if (highlightedColorize == Color.TRANSPARENT) null else highlightedColorize
        }
        get() = stateDrawable.pressedColorize

    var disabledColorize: Int?
        set(disabledColorize) {
            stateDrawable.disabledColorize = if (disabledColorize == Color.TRANSPARENT) null else disabledColorize
        }
        get() = stateDrawable.disabledColorize


    // --
    // Interaction
    // --

    override fun onClick(view: View?) {
        tapEvent?.let {
            eventObserver?.observedEvent(it, this)
        }
    }

}
